### PR TITLE
Array sort

### PR DESCRIPTION
**Pull request by okazkayasi**
Implemented https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

#599

### CODE REVIEW

**Comment by okazkayasi** [comment]
@microsoft-github-policy-service agree

File: packages/core/src/array.ts [hunk 1]
```diff
@@ -169,6 +169,24 @@ Array.prototype.reduce = function (callbackfn: any, initialValue: any) {
     return initialValue
 }
 
+Array.prototype.sort = function <T>(compareFn?: (a: T, b: T) => number) {
+    if (!compareFn) {
+        compareFn = (a: any, b: any) => a - b
```
Comment (mmoskal) [inline packages/core/src/array.ts:174]: The default JS comparison function compares strings. For example:

```
> ["9", "80", "1"].sort()
[ '1', '80', '9' ]
> [9, 80, 1].sort()
[ 1, 80, 9 ]
> ["b", "z", "a"].sort()
[ 'a', 'b', 'z' ]
>
```

So I guess it should be sth like:

```
a=a + ""
b=b + ""
if (a < b) return -1;
else if (a > b) return 1;
else return 0;
```

File: devs/run-tests/24arraymap.ts [hunk 2]
```diff
@@ -152,6 +152,56 @@ function testArrayAt() {
     assert(str.at(-6) === undefined, "arrayAtUndefined")
 }
 
+function localeCompareHelper(a: string, b: string) {
+    let i = 0
```
Comment (mmoskal) [inline devs/run-tests/24arraymap.ts:156]: this is exactly what `<` and `>` do on strings, so the function can be simplified to:

```ts
if (a < b) return -1;
else if (a > b) return 1;
else return 0;
```

File: packages/core/src/corelib.d.ts [hunk 3]
```diff
@@ -404,6 +404,18 @@ interface Array<T> {
      * @param separator A string used to separate one element of the array from the next in the resulting string. If omitted, the array elements are separated with a comma.
      */
     join(separator?: string): string
+
+    /**
+     * Sorts an array in place.
+     * This method mutates the array and returns a reference to the same array.
```
Comment (mmoskal) [inline packages/core/src/corelib.d.ts:410]: Maybe add a comment about using insertion sort (n^2 complexity) ?

**Review by mmoskal — CHANGES_REQUESTED** [review]
Thank you, this is great!

Just a few small nits.

File: devs/run-tests/24arraymap.ts (hunk 2 above)
Comment (okazkayasi) [reply devs/run-tests/24arraymap.ts]: Changed the implementation of `localeCompareHelper`

File: packages/core/src/corelib.d.ts (hunk 3 above)
Comment (okazkayasi) [reply packages/core/src/corelib.d.ts:414]: Added a comment that explains the method and complexity

File: packages/core/src/array.ts (hunk 1 above)
Comment (okazkayasi) [reply packages/core/src/array.ts]: Thanks for the review. I made the changes including this one @mmoskal

File: devs/run-tests/24arraymap.ts [hunk 4]
```diff
@@ -152,6 +152,49 @@ function testArrayAt() {
...
+        { name: "Bob", age: 25 },
+        { name: "Charlie", age: 35 },
+    ]
+    people.sort((a, b) => a.age - b.age)
+    assert(people.map(p => p.name).join() === "Bob,Alice,Charlie", "sort4")
+
+    const mixed = [5, "apple", 2, "banana", 9, "cherry"]
+    mixed.sort((a, b) => {
+        if (typeof a === "string" && typeof b === "string") {
+            localeCompareHelper(a, b)
```
Comment (mmoskal) [inline devs/run-tests/24arraymap.ts:186]: `return` missing? not sure why this isn't an error...

**Review by mmoskal — CHANGES_REQUESTED** [review]
One small change ;)

File: devs/run-tests/24arraymap.ts (hunk 4 above)
Comment (okazkayasi) [reply devs/run-tests/24arraymap.ts]: It's an error in the tests. Not sure how I missed that. Maybe deleted after running the tests. 

Added the return now. Thank you.

**Review by mmoskal — APPROVED** [review]

**Comment by mmoskal** [comment]
thank you very much!

**Pull request merged**